### PR TITLE
Use dof_indices_contiguous_sm in evaluation functions

### PR DESCRIPTION
**Pull request by peterrum**
Depends on #11074 and #11092.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -4945,18 +4951,29 @@ namespace internal
...
-    (void)vec;
-    return nullptr;
+    // note: no hp is supported
+    if (is_valid_mode_for_sm &&
+        dof_info->dof_indices_contiguous_sm[0 /*any index (<3) should work*/]
+            .size() > 0 &&
+        active_fe_index == 0)
+      return &vec.shared_vector_data();
+    else
+      return nullptr;
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_evaluation.h:4966]: I am not sure about this criteria to enable or not SM in `FE(Face)Evaluation`

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -5020,8 +5044,13 @@ inline void
 FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
   read_dof_values(const VectorType &src, const unsigned int first_index)
 {
-  const auto src_data =
-    internal::get_vector_data<n_components_>(src, first_index);
+  const auto src_data = internal::get_vector_data<n_components_>(
+    src,
+    first_index,
+    this->dof_access_index ==
+      internal::MatrixFreeFunctions::DoFInfo::dof_access_cell,
```
Comment (peterrum) [inline include/deal.II/matrix_free/fe_evaluation.h:5051]: ... this is the second part of the criteria.

**Comment by peterrum** [comment]
I have rebased this PR.

/rebuild

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 3]
```diff
@@ -4670,25 +4670,29 @@ FEEvaluationBase<dim, n_components_, Number, is_face, VectorizedArrayType>::
...
-        (void)comp;
-
-        /*
         for (unsigned int v = 0; v < n_filled_lanes; ++v)
           {
+            Assert(ind < this->dof_info->dof_indices_contiguous_sm.size(),
+                   ExcNotImplemented());
+            Assert(this->cell * VectorizedArrayType::size() + v <
+                     this->dof_info->dof_indices_contiguous_sm[ind].size(),
+                   ExcNotImplemented());
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h:4679]: Are these really not-implemented cases or rather out of range, i.e. `AssertIndexRange()`?

**Review by kronbichler — APPROVED** [review]
Looks good.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h:4679]: Don't blame me. Blame `github-linux / linux debug cuda-10`:

```
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/fe_evaluation.h: In lambda function:
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/fe_evaluation.h:4675:187: error: base operand of ‘->’ is not a pointer
             AssertIndexRange(ind,
                                                                                                                                                                                           ^ 
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/fe_evaluation.h:4675:187: error: base operand of ‘->’ is not a pointer
/home/runner/work/dealii/dealii/include/deal.II/matrix_free/fe_evaluation.h:4675:236: error: template argument 2 is invalid
             AssertIndexRange(ind,
```

See the second commit^^

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h:4679]: Does it work if you put the arguments into additional braces?

Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: Let's see.

Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: Any other ideas?

Comment (kronbichler) [reply include/deal.II/matrix_free/fe_evaluation.h]: How about writing out the check, i.e.,
```
   Assert(ind < this->dof_info->dof_indices_contiguous_sm.size(),
              ExcIndexRange(ind, 0, this->dof_info->dof_indices_contiguous_sm.size()));
```
In any case, an annoying error.

Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: Next round...

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
Looks good now.

File: include/deal.II/matrix_free/fe_evaluation.h (hunk 3 above)
Comment (peterrum) [reply include/deal.II/matrix_free/fe_evaluation.h]: OK. Then let me squash the commits.

**Pull request merged**